File: example-java/src/main/java/dev/b3nedikt/restring/example/SampleStringsLoader.java
package dev.b3nedikt.restring.example;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import dev.b3nedikt.app_locale.AppLocale;
import dev.b3nedikt.restring.PluralKeyword;
import dev.b3nedikt.restring.Restring;

/**
 * A sample of a {@link Restring.StringsLoader}. In a real application,
 * you might call an API to get your strings.
 * <p>
 * All overridden methods will be called on background thread.
 */
public class SampleStringsLoader implements Restring.StringsLoader {

    @NotNull
    @Override
    public List<Locale> getLocales() {
        return AppLocale.getSupportedLocales();
    }

    @NotNull
    @Override
    public Map<String, CharSequence> getStrings(@NotNull Locale locale) {
        final Map<String, CharSequence> map = new HashMap<>();
        if (locale == Locale.ENGLISH) {
            map.put("title", "Title (from restring).");
            map.put("subtitle", "Subtitle (from restring).");
        } else if (locale == Locale.US) {
            map.put("title", "Title US (from restring).");
            map.put("subtitle", "Subtitle US (from restring).");
        } else if (locale == Locales.LOCALE_AUSTRIAN_GERMAN) {
            map.put("title", "Titel (aus restring).");
            map.put("subtitle", "Untertitel (aus restring).");
        }
        return map;
    }

    @NotNull
    @Override
    public Map<String, Map<PluralKeyword, CharSequence>> getQuantityStrings(@NotNull Locale locale) {
        final Map<String, Map<PluralKeyword, CharSequence>> map = new HashMap<>();

        final Map<PluralKeyword, CharSequence> quantityStrings = new HashMap<>();
        quantityStrings.put(PluralKeyword.ONE, "%d quantity string " + locale + " (from restring)");
        quantityStrings.put(PluralKeyword.OTHER, "%d quantity strings " + locale + " (from restring)");

        map.put("quantity_string", quantityStrings);
        return map;
    }

    @NotNull
    @Override
    public Map<String, CharSequence[]> getStringArrays(@NotNull Locale locale) {
        final Map<String, CharSequence[]> map = new HashMap<>();

        final CharSequence[] stringArray = new CharSequence[2];
        stringArray[0] = "String Array 1 " + locale + " (from restring)";
        stringArray[1] = "String Array 2 " + locale + " (from restring)";

        map.put("string_array", stringArray);
        return map;
    }
}

File: example/src/main/java/dev/b3nedikt/restring/example/SampleStringsLoader.kt
package dev.b3nedikt.restring.example

import dev.b3nedikt.app_locale.AppLocale
import dev.b3nedikt.restring.PluralKeyword
import dev.b3nedikt.restring.Restring
import dev.b3nedikt.restring.example.Locales.LOCALE_AUSTRIAN_GERMAN
import java.util.*

/**
 * A example of a [Restring.StringsLoader].
 * In real applications, you might call an API to get your strings.
 */
object SampleStringsLoader : Restring.StringsLoader {

    override val locales = AppLocale.supportedLocales

    override fun getStrings(locale: Locale): Map<String, CharSequence> {
        val map = mutableMapOf<String, CharSequence>()
        when (locale) {
            Locale.ENGLISH -> {
                map["title"] = "Title (from restring)."
                map["subtitle"] = "Subtitle (from restring)."
                map["a_string_not_in_strings_xml"] = "A string resource not in strings xml $locale"
            }
            Locale.US -> {
                map["title"] = "Title US (from restring)."
                map["subtitle"] = "Subtitle US (from restring)."
                map["a_string_not_in_strings_xml"] = "A string resource not in strings xml $locale"
            }
            LOCALE_AUSTRIAN_GERMAN -> {
                map["title"] = "Titel (aus restring)."
                map["subtitle"] = "Untertitel (aus restring)."
                map["a_string_not_in_strings_xml"] = "A string resource not in strings xml $locale"
            }
        }
        return map
    }

    override fun getStringArrays(locale: Locale): Map<String, Array<CharSequence>> {
        return mapOf(
                "string_array"
                        to
                        arrayOf<CharSequence>("String Array 1 $locale (from restring)",
                                "String Array 2 $locale (from restring)")
        )
    }

    override fun getQuantityStrings(locale: Locale): Map<String, Map<PluralKeyword, CharSequence>> {
        return mapOf(
                "quantity_string"
                        to
                        mapOf(PluralKeyword.ONE to "%d quantity string $locale (from restring)",
                                PluralKeyword.OTHER to "%d quantity strings $locale (from restring)")
        )
    }
}